src/vsr/grid.zig

```diff
@@ -571,117 +661,112 @@ pub fn GridType(comptime Storage: type) type {
-            assert(!grid.superblock.free_set.is_free(address));
-            assert(grid.writing(address, null) != .init);
-
-            assert(address > 0);
-            assert(block_type != .reserved);
-
-            read.* = .{
-                .callback = callback,
-                .address = address,
-                .checksum = checksum,
-                .block_type = block_type,
-                .read_cache = false,
-                .checkpoint_id = grid.superblock.working.checkpoint_id(),
-                .grid = grid,
-            };
-
-            grid.read_block(read);
-        }
-
-        fn read_block(grid: *Grid, read: *Grid.Read) void {
             assert(grid.superblock.opened);
             assert(grid.canceling == null);
-            assert(!grid.superblock.free_set.is_free(read.address));
-            assert(grid.writing(read.address, null) != .init);
+            if (read.coherent) {
+                assert(!grid.superblock.free_set.is_free(read.address));
+                assert(grid.writing(read.address, null) != .create);
+            }
 
             assert(read.address > 0);
-            assert(read.block_type != .reserved);
 
             // Check if a read is already processing/recovering and merge with it.
             for ([_]*const FIFO(Read){
                 &grid.read_queue,
                 &grid.read_faulty_queue,
             }) |queue| {
+                // Don't remote-repair repairs – the block may not belong in our current checkpoint.
+                if (read.callback == .from_local) {
+                    if (queue == &grid.read_faulty_queue) continue;
+                }
+
                 var it = queue.peek();
                 while (it) |queued_read| : (it = queued_read.next) {
                     if (queued_read.address == read.address) {
-                        assert(queued_read.checksum == read.checksum);
-                        assert(queued_read.block_type == read.block_type);
-                        queued_read.resolves.push(&read.pending);
-                        return;
+                        // TODO check all read options match
+                        if (queued_read.checksum == read.checksum) {
+                            queued_read.resolves.push(&read.pending);
+                            return;
+                        } else {
+                            assert(!queued_read.coherent or !read.coherent);
+                        }
                     }
                 }
             }
 
-            // When Read.read_cache is set, the caller of read_block() is responsible for calling
+            // When Read.cache_read is set, the caller of read_block() is responsible for calling
             // us via next_tick().
-            if (read.read_cache) {
-                if (grid.read_block_from_cache(read.address, read.checksum)) |cache_block| {
-                    grid.read_block_resolve(read, cache_block);
+            if (read.cache_read) {
```

Comment:
Heh, that's some fully control flow here. I was wondering if maybe we could push the loop above to the synchronous path, so that this callback has to deal just with the cache, but that doesn't work. 

I guess this could be a bit better if we learn to pin data in cache -- that way we can synchronously decide whether we need to read from cache or from disk, simplifying the call-site.